api/app/Http/Controllers/EntityController.php

```diff
@@ -321,6 +326,99 @@ protected function searchAllEntities($queryString, $limit = null, $offset = null
     }
 
     /**
+     * @param $query
+     * @param null $totalCount
+     * @return \Elasticquent\ElasticquentResultCollection
+     */
+    protected function complexSearchAllEntities($query, &$totalCount = null)
+    {
+        /* @var ElasticquentTrait $model */
+        $model = $this->getModel();
+
+        $searchResults = $model->complexSearch([
+            'index' => $model->getIndexName(),
+            'type' => $model->getTypeName(),
+            'body' => $this->translateQuery($query),
+        ]);
+
+        if ($searchResults->totalHits() === 0) {
+            throw new NotFoundHttpException(sprintf('No results found for model `%s`', get_class($model)));
+        }
+
+        if (isset($totalCount) && $searchResults->totalHits() < $totalCount) {
+            $totalCount = $searchResults->totalHits();
+        }
+
+        return $searchResults;
+    }
+
+    /**
+     * Takes a query and translates it into a query that elastic search understands.
+     *
+     * Expect query to be in form, e.g.:
+     * {
+     *      _all: [ "stringA", "stringB" ],
+     *      someField: [ "stringA" ],
+     *      _nestedEntity: { key: [ "stringA", "stringB" ]
+     * }
+     *
+     * Notes:
+     * - Empty values will be removed from search completely.
+     * - You must pass an array, even if it only contains 1 string.
+     *
+     * @param $query
+     * @return mixed
+     */
+    private function translateQuery($query)
+    {
+        $processedQuery['query']['bool']['must'] = [];
+
+        foreach ($query as $key => $value) {
+            if (substr($key, 0, 1) == '_' && $key != '_all') { // Nested entity
+                reset($value);
+                $fieldKey = key($value);
+
+                foreach ($value[$fieldKey] as $fieldValue) {
+                    if (! empty($fieldValue)) {
+                        array_push($processedQuery['query']['bool']['must'],
+                            [
+                                'nested' => [
+                                    'path' => snake_case(substr($key, 1)),
```

Comment:
This snake_case call is being repeated for every iteration of the loop (3 times) when it can be moved to a single variable.